implementations/rust/ockam/ockam_command/src/state.rs

```diff
@@ -0,0 +1,535 @@
+        let identity = match &config.identity {
+            Some(identity) => self.identities.get(identity)?,
+            None => self.identities.default()?,
+        };
+        self.nodes.create(vault, identity, name)
+    }
+
+    pub fn node(&self, name: &str) -> anyhow::Result<NodeState> {
+        let vault = {
+            let name = self.nodes.vault_name(name)?;
+            self.vaults.get(&name)?
+        };
+        let identity = {
+            let name = self.nodes.identity_name(name)?;
+            self.identities.get(&name)?
+        };
+        self.nodes.get(vault, identity, name)
+    }
+}
+
+pub struct VaultsState {
+    dir: PathBuf,
+}
+
+impl VaultsState {
+    fn new(cli_path: &Path) -> anyhow::Result<Self> {
+        let dir = cli_path.join("vaults");
+        std::fs::create_dir_all(&dir)?;
+        Ok(Self { dir })
+    }
+
+    pub fn create(&self, name: &str, config: VaultConfig) -> anyhow::Result<VaultState> {
+        let path = {
+            let mut path = self.dir.clone();
+            path.push(format!("{}.json", name));
+            path
+        };
+        let contents = serde_json::to_string(&config)?;
+        std::fs::write(&path, contents)?;
+        Ok(VaultState { path, config })
+    }
+
+    pub fn get(&self, name: &str) -> anyhow::Result<VaultState> {
+        let path = {
+            let mut path = self.dir.clone();
+            path.push(format!("{}.json", name));
+            if !path.exists() {
+                return Err(anyhow::anyhow!("vault `{name}` does not exist"));
+            }
+            path
+        };
+        let contents = std::fs::read_to_string(&path)?;
+        let config = serde_json::from_str(&contents)?;
+        Ok(VaultState { path, config })
+    }
+
+    pub fn default(&self) -> anyhow::Result<VaultState> {
+        let path = {
+            let mut path = self.dir.clone();
+            path.push("default");
+            std::fs::canonicalize(&path)?
+        };
+        let contents = std::fs::read_to_string(&path)?;
+        let config = serde_json::from_str(&contents)?;
+        Ok(VaultState { path, config })
+    }
+
+    pub fn set_default(&self, name: &str) -> anyhow::Result<VaultState> {
+        let original = {
+            let mut path = self.dir.clone();
+            path.push(format!("{}.json", name));
+            path
+        };
+        let link = {
+            let mut path = self.dir.clone();
+            path.push("default");
+            path
+        };
+        std::os::unix::fs::symlink(&original, &link)?;
+        let contents = std::fs::read_to_string(&original)?;
+        let config = serde_json::from_str(&contents)?;
+        Ok(VaultState {
+            path: original,
+            config,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct VaultState {
+    path: PathBuf,
+    pub config: VaultConfig,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
+#[serde(untagged)]
+pub enum VaultConfig {
```

Comment:
I'm not entirely sure how things are structured, but have the _feeling_ this should be somewhere else.
The reasoning is

1) Create a node `$ockam node create n1`
2) Start a secure channel with a _specific_  vault and identity:  `$ockam [...]  --identity id1  --vault v1`  

The question is how the cli,  running on step 2,   pass the vault and identity to use to `n1` (the running background node) .   I _think_  serializing the entire change history is not the best approach as that's something that can change  (like rotating the keys on the identity).   So perhaps passing the path to the file is better?. 

What do you think?   I don't have this part particularly clear. 

(also,  I'm assuming the background node doesn't have knowledge of this top-level OCKAM_HOME thing,  and instead only knows _his_ particular node' state directory,   and any other path that the cli communicate to it)

Reply:
When running `ockam node create --identity id1 --vault v1 [...]` the command will look like this:

```rust
let vault_path = options.vaults.get(&vault_name).path;
let identity_path = options.identities.get(&identity_name).path;
let node_config = NodeConfig::new(&vault_path, &identity_path);
let node_state = options.nodes.create(&node_name, node_config);
``` 

The `node_state` contains the _paths_ to the vault and identity. That's what we will send to the node upon creation. Then, every time we need to use the vault or identity, we will recreate them from scratch, use them, and drop them. That way, if we need to mutate an identity (rotate the key), we can do that from a command for example, and the nodes using that identity will use the updated version.

Does this make sense to you?

Reply:
ok.  There are two path when talking  file-based vaults:    One to the file that said that the vault is a file-based one (instead of a kms based one, etc)  and that has the path to the actual file-backed vault.  And the other is that path to the file-backed vault itself.

From the command to the running node we pass the first path,  the path to vault' definition.     The background running node needs to be able to parse that file struct (what is called `VaultConfig` here).   I mean either parse it from disk,  or the client need to serialize that `VaultConfig` struct and give it to the background node.   In either case the server code needs to have access to `VaultConfig`'  definition,   that's what I'm unsure how is done.
But ok to move definitions around latter.